### ISSUE TITLE
IDE0059 on ref parameter reassignment

### ISSUE THREAD

**Issue by NextTurn** [dotnet/roslyn#37871]
**Version Used**: Visual Studio 2019 version 16.3 Preview 1

**Steps to Reproduce**:

```cs
using System.Runtime.CompilerServices;

class Program
{
    void Method(ref int param)
    {
        {
            param = ref Unsafe.Add(ref param, 1); // IDE0059
            param = default;
        }

        {
            param = ref Unsafe.Add(ref param, 1); // No warning
            param = ref Unsafe.Add(ref param, 1); // IDE0059
            param = default;
        }

        {
            param = ref Unsafe.Add(ref param, 1); // No warning
            param = ref Unsafe.Add(ref param, 1); // No warning
        }

        ref int local = ref param;

        {
            local = ref Unsafe.Add(ref param, 1); // No warning
            local = default;
        }

        {
            local = ref Unsafe.Add(ref param, 1); // No warning
        }
    }
}
```

**Expected Behavior**:

IDE0059 is not reported for the reassignment.

**Actual Behavior**:

IDE0059 is reported for the reassignment.

**Comment by mavasani** [comment]
I believe this is by design. Couple of places that flag IDE0059 (unnecessary assignment of a value) can be changed as following without changing semantics:

```csharp
   _ = ref Unsafe.Add(ref param, 1);
   param = default;
```

For example, both programs below produce the same console output:
```csharp
using System;
using System.Runtime.CompilerServices;

class Program
{
    static void Main()
    {
        int x = 0;
        Method(ref x);
    }

    static void Method(ref int param)
    {
        param = ref Unsafe.Add(ref param, 1);
        param = default;

        Console.WriteLine(param);
    }
}
```

```csharp
using System;
using System.Runtime.CompilerServices;

class Program
{
    static void Main()
    {
        int x = 0;
        Method(ref x);
    }

    static void Method(ref int param)
    {
        _ = ref Unsafe.Add(ref param, 1);
        param = default;

        Console.WriteLine(param);
    }
}
```

**Comment by NextTurn** [comment]
Then should it be considered as a bug that IDE0059 is not flagged elsewhere?

**Comment by mavasani** [comment]
Other assignments to “param” are not redundant assignments, and analyzer conservatively bails out for ref locals, though it can be possibly enhanced.

**Comment by NextTurn** [comment]
Your two methods are not equivalent.

```cs
using System;
using System.Runtime.CompilerServices;

class Program
{
    static void T1(ref int param)
    {
        param = ref Unsafe.Add(ref param, 1);
        param = default;
    }

    static void T2(ref int param)
    {
        _ = ref Unsafe.Add(ref param, 1);
        param = default;
    }

    static void Main()
    {
        {
            int[] arr1 = new[] { 1, 2 };
            T1(ref arr1[0]);
            Console.WriteLine($"arr1 : {arr1[0]}, {arr1[1]}"); // arr1 : 1, 0
        }

        {
            int[] arr2 = new[] { 1, 2 };
            T1(ref arr2[0]);
            Console.WriteLine($"arr2 : {arr2[0]}, {arr2[1]}"); // arr2 : 1, 0
        }

        {
            int[] arr3 = new[] { 1, 2 };
            ref int a = ref arr3[0];

            T1(ref a);
            Console.WriteLine($"arr3 : {arr3[0]}, {arr3[1]}"); // arr3 : 1, 0

            T2(ref a);
            Console.WriteLine($"arr3 : {arr3[0]}, {arr3[1]}"); // arr3 : 0, 0
                                                               //        ^
        }
    }
}
```

**Comment by mavasani** [comment]
@NextTurn Thanks, you are correct. I will update the analyses to be conservative in presence of ref assignments.
